### ISSUE TITLE
PDF417 anyone?

### ISSUE THREAD

**Issue by axxel** [nu-book/zxing-cpp#145]
After I'm almost done with #89 the remaining main 'offender' in terms of performance is the PDF417 decoder. I started looking at the code and found it to be pretty hard to read (to put it politically correct :)). I would also question the general approach the authors took: the code basically performs individual horizontal line-scanning operations. This can generally only work as long as the rotation of the symbol is so small that at least one complete codeword is visible when scanning a horizontal line. I tested the code by starting with a [perfect](https://github.com/nu-book/zxing-cpp/blob/master/test/samples/pdf417-3/01.png) symbol upscaled it by a factor of 2, to make it even easier, and rotated it. Even just 3° rotation are already too much. On the other hand the current code deals happily with corrupt samples like [this](https://github.com/nu-book/zxing-cpp/blob/master/test/samples/pdf417-3/11.png).

I would argue that it is more important to handle real-world rotations caused by hand-held cameras than to be able to deal with failing dot-matrix printers or thick black lines half across your code. Meaning: if a (faster) replacement would deal with the first on the expense of the second, it would be a net win?

I doubt I have the time to re-implement the PDF417 decoder but still wanted to ask for opinions on the matter, just in case.

**Comment by axxel** [comment]
Addendum: I just found a one-line change that immediately makes the code about twice as fast in the "falsepositive" use case: simply don't look for the stop-guard-pattern if no start-guard-pattern has been found. If I do that, there are 4 samples that fail as a result. 3 of those are not valid PDF417 symbols because their start-guard is missing. There is a "compact PDF417" version available but only the stop-guard can be missing, not the start-guard.

The 4th is one of the [multi-symbol examples](https://github.com/nu-book/zxing-cpp/blob/master/test/samples/pdf417-4/04-01.png). That one is actually also invalid because it does not have the required row height of at least 3X.

I propose to remove the invalid samples and take that 2x speed improvement we get 'for free'.

**Comment by huycn** [comment]
In our use cases, the change you propose would be a net win.

**Comment by axxel** [comment]
To summarize #152: aside from better support for rotated symbols, also support vertically flipped symbols.

**Comment by tailsu** [comment]
The individual barcodes in a PDF417 stack have 3 different symbologies. The one used for each row depends on the row index % 3. So every row not only encodes its data but also its index in the stack mod 3. While decoding a scanline, you should be seeing only symbols from the same symbology. If at some point during decoding you start seeing some junk and then symbols from an adjacent symbology, then you know you've entered an adjacent row.

So an improved scanner should be able to detect that it's entered an adjacent row, go back a bit, rotate the scanline up or down a bit and try to find the continuation of the current row. You'll need a function that can extract rotated pixel rows from the input image, not just horizontal. Then, you can give them to the current horizontal scan line decoder.

I can't estimate how much effort it would be to improve the current implementation to do that. It seems that there are a lot of pieces missing.

**Comment by axxel** [comment]
That is correct. I believe the intended way to decode it is by going through it scan-line by scan-line and if you cross from one symbology into another, you simply assign the code words after the garbage to the correct new places. This way you loose the code words that were garbage due to the crossing but if you have enough error correction redundancy overall, you still get a valid result. That approach allows for a 1-D line-scanning device to decode the symbol.

I thought about going a different route: first detect the 4 corners of the whole symbol and then sample it on a grid, just like the other 2D codes like the QRCode are handled as well. That approach is likely faster at runtime and also easier from the implementation perspective but likely not as robust when it comes to deformations caused by the camera lens or non-flat surfaces. That said, it should work a lot better where the symbol is rotated more than just a few degrees. I believe you can't possibly get a valid result with the line-scanning approach if the symbol is rotated 45°.

Are you interested / in a position to start working on that?

**Comment by tailsu** [comment]
Not sure yet. I'll have to think about it.

**Comment by axxel** [comment]
> You'll need to estimate the total number of rows of the barcode before applying a grid on top.

Yes. That information is stored in the first column of the symbol, iirc. So you would need to start by interpreting at least one row of the first column.

You could also take a hybrid approach: locate the symbol, then scanning it 'line by line' parallel to the top/bottom border. That could actually be an improvement step after starting with the grid sampling.

If you should start coding something, please start 'from scratch'. This part of the code base is the one furthest away from idiomatic c++ and pretty complex/opaque in its logic, too. If you are interested, we could also have a chat on the phone.

May I ask what your interest in / application for PDF417 is?

**Comment by tailsu** [comment]
I'm looking for a PDF417 reader for US driver licenses. The print size is tiny and the quality there is awful. The current PDF417 scanner fails to decode them in >75% of view angles and sizes.

If I decide to try to implement a better reader, I will most likely like to be able to use OpenCV. This is one thing that's stopping me from considering developing a pure zxing-cpp implementation.

**Comment by axxel** [comment]
I see. We will definitively not add code with a OpenCV dependency to the project but I can imagine that OpenCV can help a lot in preprocessing/locating the symbol.

For reference and a potential future endeavor of my own, could you provide a representative sample of your failing input or even better, provide a link to a whole bunch of samples stored somewhere?

**Comment by tailsu** [comment]
> For reference and a potential future endeavor of my own, could you provide a representative sample of your failing input or even better, provide a link to a whole bunch of samples stored somewhere?

This fellow was kind enough to post one of his own. https://github.com/nu-book/zxing-cpp/issues/152
It's a South-African driver license, but the US are rather similar. They seem to be printed in a similar quality and module and row size, but contain variable amount of information depending on the issuing state.

Unfortunately, the few driver licenses that I have are under an NDA. They contain personally identifiable information after all. I couldn't just slap them here in the issue.

**Comment by tailsu** [comment]
Just FYI regarding US driver licenses. The data in the PDF417 barcode follows the AAMVA DL/ID spec. I've seen generators available online but I haven't tried them. I've also worked with this [parser library](https://github.com/c0shea/IdParser) for the format. It converts the raw encoded data in the barcode into a structured format. It's in C#

**Comment by axxel** [comment]
@tailsu Have you started working on a re-implementation and could share your progress? I'm thinking about starting to look into the code now and thought asking first might save me some time :)

**Comment by tailsu** [comment]
@axxel Yes, I made some progress.

I actually went with your idea to derive the perspective transform of the barcode and then read out the data using the transformed coordinate frame. I managed to augment `GetModuleBitCount` and the entire call chain to it to apply a `PerspectiveTransform` while iterating over the image, basically the same way it is done in `GridSampler.cpp`. This works remarkably well, especially for rotated barcodes, but only if the initial bounds estimate is accurate - which it very often isn't.

In fact, the current implementation of `PDFScanningDecoder` can often decode the barcode properly even if only a part of the left and right guards were found. That, of course, will never work if you feed it an incorrectly estimated perspective transform. So I had to debug and basically rewrite half of `PDFDetector` to make the initial position estimate accurate enough.

I wanted to reuse as much of the current implementation as possible. That's why I didn't go ahead with a clean reimplementation. That said, I did use OpenCV for the tricky bits in `PDFDetector`, so my code can't be integrated back upstream (and my company won't allow it anyway), but at least you can reuse my insights.

**Comment by axxel** [comment]
Thanks for the feedback. I started looking into the code base more closely and am more convinced then ever that I'd prefer to re-implement it :). I'll start with committing a 'pure' version of the detector, which will of course only help in very limited situations.

I'm still thinking about the best approach ('simulate a line scanner' vs. the 'perspective transform from corner points'). Both have their pros and cons. So your new `PDFDetector` can deal with input like [10.png](https://github.com/nu-book/zxing-cpp/blob/master/test/samples/pdf417-3/10.png), [11.png](https://github.com/nu-book/zxing-cpp/blob/master/test/samples/pdf417-3/11.png) and [12.png](https://github.com/nu-book/zxing-cpp/blob/master/test/samples/pdf417-3/12.png)? I wonder whether the 10.png example is actually an interesting use case or whether support for arbitrary rotations is way more important in practice?

A comment about your company's attitude: it sucks ;). While they have the license to do so, I'd appreciate if they considered to allow you bring back any improvements to the core functionality of the library, while I support the idea of building something completely proprietary on top of that. That is how I personally started out contributing to this project, because I needed a usable DataMatrix detector for my own company. That said, I'm still not interested in adding an OpenCV dependency. ;)

**Comment by tailsu** [comment]
> A comment about your company's attitude

**Comment by tailsu** [comment]
> So your new PDFDetector can deal with input like 10.png, 11.png and 12.png?
That's a good question, I will need to test it. I'm confident that 11 will work, not sure about the rest. I'll keep you posted.

**Comment by axxel** [comment]
As I'm further looking into this I have a question: Did you keep the `GetModuleBitCount` approach of scanning through individual pixels and counting the number of successive white/black ones or did you sample each code word character at 17 positions based on your perspective transform?

**Comment by tailsu** [comment]
@axxel I kept the bit-counting approach. Since the whole barcode is basically a stack of 1D barcodes, I thought that the best way to decode each codeword eventually might be to use PatternMatchVariance instead of the current algorithm. For that you'd need the counts.

But I'm not really sure if migrating to PatternMatchVariance is even viable.

**Comment by tailsu** [comment]
@axxel I tested the three images and all were recognized correctly. I had no doubt that 11 would work in the current implementation. In 10 my algorithm estimates the top-right corner incorrectly, about 1.5 rows higher than it should be, but that was apparently not a big deal. I guess that the projection error is strongest in the places where the codewords are destroyed with the marker anyway.

The distortion is 12 is not projective, but it's just small enough to still work.

**Comment by jkflying** [comment]
As discussed in #295 , I have some performance issues (quality, not speed) with the PDF417 decoder and I'd be happy to help trying to improve it. Considering that in my case we are detecting the barcode but getting checksum errors, is there anything I can do to help that doesn't require rewriting the whole thing? Maybe the thresholding is sub-optimal? Or maybe it needs to be upscaled because the binarization is having rounding issues? Or there are alignment errors due to image distortion?

Are there any tools we can use to dump the binary images and compare / overlay them on the input images? I guess between binarization and perspective distortion correction, it's the biggest place for performance improvement?

I'm a computer vision engineer, so that's the easiest place for me to improve things. Also AFAIK once it's past the 'fuzzy' data the rest of it should be deterministic, so even if the code is hard to understand it should be 'right'. It's not like we aren't applying the ECC or something else which would be an easy gain.

Anyway, let me know where I could help and if there are any tools to make life easier :+1:

**Comment by axxel** [comment]
a) in terms of 'tools', you might look at the `LogMatrixWriter` class, which is e.g. used inside `QRReader` for debugging purposes.
b) your idea of simply up-scaling the image beforehand could indeed do the trick.

As I said before, the main issue in your example is lack of resolution in my opinion. I don't see any distortion/rotation/etc. In that sense, it is actually almost a `pure` symbol. Some kind of up-scaling maybe with additional edge enhancing / pre-binarization will definitively make this example work (if the symbol itself is not corrupt, which you said it was not).

**Comment by jkflying** [comment]
Interesting, upscaling 2x with Lanczos was sufficient to get my sample to read properly. Of course it's probably pretty marginal, I'll also look at sharpening. I guess there's no value in pre-binarizing using a robust threshold like OTSU, zxing-cpp already has good binarization?

**Comment by axxel** [comment]
The implemented binarization is mostly working but has it's limits. It is not as mathematically sound as OTSU but tries to achieve approximately the same thing. The most effect in terms of improvement could be had with a pre-applied classifier to distinguish between areas that are likely barcode and those that are not. Then limit the input of the binarization to the respective areas thereby allowing a local and much more precise threshold determination. I guess a full blown NN approach could do wonders here. But that is not my area of expertise. I've spent a considerable amount of time working on some more ad-hoc heuristics but that did not lead to something strictly superior to what we have right now, so this effort did not bear (visible) fruit yet. 

If you are interested and want to talk about details, maybe we can chat. In that case, drop me line via e-mail.

**Comment by niamdsamsd** [comment]
> > 在顶部应用网格之前，您需要估计条形码的总行数。
> 
> 是的。该信息存储在符号 iirc 的第一列中。因此，您需要从解释第一列的至少一行开始。
> 
> 您也可以采用混合方法：找到符号，然后平行于顶部/底部边框“逐行”扫描它。这实际上可能是开始网格采样后的一个改进步骤。
> 
> 如果你应该开始编码，请“从头开始”。代码库的这一部分与惯用的 c++ 相去甚远，其逻辑也相当复杂/不透明。如果您有兴趣，我们也可以通过电话聊天。
> 
> 请问您对PDF417的兴趣/申请是什么？
How to do ZXing C++ to identify MICRO PDF417
